AI_Project/src/vec2.cpp

```diff
@@ -0,0 +1,269 @@
+#include "vec2.h"
+#include <cmath> // for std::sqrtf
+#include <limits>// for std::numeric_limits
+
+vec2::vec2()
+  :x(0.0f),
+  y(0.0f)
+{}
+
+vec2::vec2(float const& X,
+           float const& Y)
+  : x(X),
+  y(Y)
+{}
+
+vec2
+vec2::operator+(vec2 const& other) const
+{
+  return add(other);
+}
+
+vec2&
+vec2::operator+=(vec2 const& other)
+{
+  return addToSelf(other);
+}
+
+vec2
+vec2::operator-(vec2 const& other) const
+{
+  return subtract(other);
+}
+
+vec2&
+vec2::operator-=(vec2 const& other)
+{
+  return  subtractFromSelf(other);
+}
+
+vec2
+vec2::operator*(float const& scalar) const
+{
+  return mulScalar(scalar);
+}
+
+vec2&
+vec2::operator*=(float const& scalar)
+{
+  return mulSelfByScalar(scalar);
+}
+
+float
+vec2::operator*(vec2 const& other) const
+{
+  return dot(other);
+}
+
+bool
+vec2::operator==(vec2 const& other) const
+{
+  //controls how many digits of accuracy we check for
+//  static constexpr int digitsBeforeTheDot = 6; // Threshold
+  static constexpr float const minimalThreshold = 1.0f / 1000.0f;// 0.001
+  vec2 const delta = this->subtract(other);
+
+  if( std::fabsf(delta.x) <= minimalThreshold &&
+     std::fabsf(delta.y) <= minimalThreshold )
+  {
+    return true;
+  }
+
+  //float const scaleEpsilonInX =
+  //  std::numeric_limits<float>::epsilon() * std::fabsf(Sum.x);
+
+  //float const scaleEpsilonInY =
+  //  std::numeric_limits<float>::epsilon() * std::fabsf(Sum.y);
+
+  //if( std::fabsf(delta.x) <= (scaleEpsilonInX * digitsBeforeTheDot) &&
+  //   std::fabsf(delta.y) <= (scaleEpsilonInY * digitsBeforeTheDot) )
+  //{
+  //  return true;
+  //}
+
+  //else if( std::fabsf(delta.x) < std::numeric_limits<float>::min() &&
+  //       std::fabsf(delta.y) < std::numeric_limits<float>::min() )
+  //{
+  //  return true;
+  //}
+      
+  return false;
+}
+
+bool 
+vec2::operator!=(vec2 const& other) const
+{
+  if(*this == other)
+    return false;
+      
+  return true;
+}
+
+
+bool 
+vec2::operator<(vec2 const& other) const
+{
+  if( this->magnitudeSqr() < other.magnitudeSqr() )
+    return true;
+
+  return false;
+}
+
+bool 
+vec2::operator>(vec2 const& other) const
+{
+
+  if( this->magnitudeSqr() > other.magnitudeSqr() )
+    return true;
+
+  return false;
+}
+
+vec2
+vec2::add(vec2 const& other) const
+{
```

Comment:
No usar const y this en la misma función
Evite usar el mismo nombre del miembro para parámetros